Zend/zend_enum.c

```diff
@@ -371,3 +376,128 @@ void zend_enum_register_props(zend_class_entry *ce)
+	ZEND_FE_END
+};
+
+ZEND_API zend_class_entry *zend_register_internal_enum(const char *name)
+{
+	return zend_register_internal_backed_enum_ex(name, IS_UNDEF, NULL);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_enum_ex(const char *name, const zend_function_entry *functions)
+{
+	return zend_register_internal_backed_enum_ex(name, IS_UNDEF, functions);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_backed_enum(const char *name, uint32_t backing_type)
+{
+	return zend_register_internal_backed_enum_ex(name, backing_type, NULL);
+}
+
+ZEND_API zend_class_entry *zend_register_internal_backed_enum_ex(const char *name, uint32_t backing_type, const zend_function_entry *functions)
+{
+	ZEND_ASSERT(backing_type == IS_UNDEF || backing_type == IS_STRING || backing_type == IS_LONG);
+
+	zend_class_entry ce_val, *ce;
+	INIT_CLASS_ENTRY_EX(ce_val, name, strlen(name), functions);
+	ce = zend_register_internal_class(&ce_val);
+
+	ce->ce_flags |= ZEND_ACC_ENUM;
+	ce->enum_backing_type = backing_type;
+	ce->backed_enum_table = malloc(sizeof(HashTable));
+	zend_hash_init(ce->backed_enum_table, 0, NULL, ZVAL_PTR_DTOR, 1);
+	zend_enum_register_props(ce);
+
+	zend_register_functions(ce, simple_internal_methods, &ce->function_table, EG(current_module)->type);
+	zend_do_implement_interface(ce, zend_ce_unit_enum);
+	if (backing_type != IS_UNDEF) {
+		zend_register_functions(ce, backed_internal_methods, &ce->function_table, EG(current_module)->type);
+		zend_do_implement_interface(ce, zend_ce_backed_enum);
+	}
+
+	return ce;
+}
+
+static zend_object *add_internal_enum_case(zend_class_entry *ce, zend_string *name, zval *value)
+{
+	zend_object *case_obj = zend_objects_new_persistent(ce);
+	enum_obj_init(case_obj, name, value);
+
+	zval casezv;
+	ZVAL_OBJ(&casezv, case_obj);
+	zend_class_constant *c = zend_declare_class_constant_ex(ce, name, &casezv, ZEND_ACC_PUBLIC, NULL);
+	ZEND_CLASS_CONST_FLAGS(c) |= ZEND_CLASS_CONST_IS_CASE;
+
+	zend_string_release(name);
+	if (value) {
+		zval_internal_ptr_dtor(value);
+	}
+
+	return case_obj;
+}
+
+ZEND_API zend_object *zend_add_enum_case_ex(zend_class_entry *ce, zend_string *name)
+{
+	ZEND_ASSERT(ce->enum_backing_type == IS_UNDEF);
+
+	name = zend_new_interned_string(name);
+
+	zend_object *case_obj = add_internal_enum_case(ce, name, NULL);
+
+	zend_string_release(name);
+
+	return case_obj;
+}
+
+ZEND_API zend_object *zend_add_enum_case_long_ex(zend_class_entry *ce, zend_string *name, zend_long value)
+{
+	ZEND_ASSERT(ce->enum_backing_type == IS_LONG);
+
+	name = zend_new_interned_string(name);
+
+	zval namezv;
+	ZVAL_STR_COPY(&namezv, name);
+	if (!zend_hash_index_add(ce->backed_enum_table, value, &namezv)) {
+		ZEND_ASSERT(0 && "Cannot add two enum values with the same backing value");
+	}
+
+	zval zv;
+	ZVAL_LONG(&zv, value);
+	zend_object *case_obj = add_internal_enum_case(ce, name, &zv);
+
+	zend_string_release(name);
```

Comment:
Same comments on ZVAL_INTERNED_STR and not needing zval_string_released on zend_new_interned_string. It seems like both zend_string_init_interned_permanent and zend_new_interned_string_request return interned strings where refcounting or freeing isn't necessary.

The code works but it's longer than it needs to be?